docs/spec/consensus/light-client.md

```diff
@@ -1,113 +1,145 @@
-  nextValSet = Validators(signedHeader.Header.Height)
-  assertThat Hash(nextValSet) == signedHeader.Header.ValidatorsHash
-  valSet = nextValSet.Validators
-  valSetHash = signedHeader.Header.ValidatorsHash
-  valSetNumber = signedHeader.ValSetNumber
-  valSetTime = nextValSet.ValSetTime
-
-votingPower(commit Commit):
-  votingPower = 0
-  for each precommit in commit.Precommits do:
-    if precommit.ValidatorAddress is in valSet and signature of the precommit verifies then
-      votingPower += valSet[precommit.ValidatorAddress].VotingPower
-  return votingPower
-
-votingPower(validatorSet []Validator):
-  for each validator in validatorSet do:
-    votingPower += validator.VotingPower
-  return votingPower
-
-updateValidatorSet(valSetNumberOfTheHeader):
-  while valSetNumber != valSetNumberOfTheHeader do
-    signedHeader = LastHeader(valSetNumber)
-    if isValid(signedHeader) then
-      setValidatorSet(signedHeader)
-    else return error
-  return
-```
+VerifyAndUpdate(state LiteClientState, height int64): error {
+  if signedHeader.Header.Height <= height then return error
+  
+  // check if certification period of the current validator set is still secure
+  if state.ValSetTime + LITE_CLIENT_CERTIFICATE_PERIOD > Now() - BFT_TIME_DRIFT_BOUND then return error
+
+  for i = state.Height+1; i < height; i++ {
+    signedHeader = Commit(i)
+    if votingPower(signedHeader.Commit) <= 2/3 totalVotingPower(state.ValSet) then return error
+    if signedHeader.Header.ValidatorsHash != state.ValSetHash then return error
+
+    nextValSetHash = signedHeader.Header.NextValidatorsHash
+
+    nextValSet = Validators(i+1).Validators
+    if Hash(nextValSet) != nextValSetHash then return (0, error)
+
+    // at this point we can install new valset
+    state.ValSet = nextValSet
+    state.ValSetHash = nextValSetHash
+    state.Height = i
+    state.ValSetTime = signedHeader.Header.Time
+  }  
+}
+
+Note that in the logic above we assume that the lite client will always go upward with respect to header verifications,
+i.e., that it will always be used to verify more recent headers. Going backward is problematic as having trust in 
+some validator set at height h does not give us trust in validator set before height h. Therefore, if lite client query
+is about smaller height, trusted validator set for smaller heights is needed. 
 
-Note that in the logic above we assume that the light client will always go upward with respect to header verifications,
-i.e., that it will always be used to verify more recent headers. In case a light client needs to be used to verify older
-headers (go backward) the same mechanisms and similar logic can be used. In case a call to the FullNode or subsequent
-checks fail, a light client need to implement some recovery strategy, for example connecting to other FullNode.
+In case a call to the FullNode or subsequent checks fail, a light client need to implement some recovery strategy, for example connecting to other FullNode.
```

Comment:
The light client should always try to use a randomized load balancing strategy, considering the possibility of malevolent eclipse attacks (or just innocuous but inconvenient stale full node data).